README.md
# NgxCanvas

<!-- Badges section here. -->
<p align="center">
  <a href="https://travis-ci.com/"><img src="https://travis-ci.com/lanxuexing/ngx-canvas.svg?branch=main" alt="Build Status"></a>
  <a href="https://npmcharts.com/compare/ngx-canvas?minimal=true"><img src="https://img.shields.io/npm/dm/ngx-canvas.svg?sanitize=true" alt="Downloads"></a>
  <a href="https://www.npmjs.com/package/ngx-canvas"><img src="https://img.shields.io/npm/v/ngx-canvas.svg?sanitize=true" alt="Version"></a>
  <a href="https://www.npmjs.com/package/ngx-canvas"><img src="https://img.shields.io/npm/l/ngx-canvas.svg?sanitize=true" alt="License"></a>
</p>

## Table of contents
- [Quick example](#quick-example)
- [Installation](#installation)
- [Usage](#usage)
- [API](#api)
- [Events](#events)
- [Types](#types)
- [Demo](#demo)

## Quick example

```html
<ngx-canvas
  width="375"
  height="667"
  [options]="options"
  (drawComplete)="drawComplete($event)"
></ngx-canvas>
```

## Installation

To install `ngx-canvas` run the following command:

```bash
npm i ngx-canvas --save
```

## Usage

First, import the `NgxCanvasModule` into the `AppModule`:

```typescript
import { NgModule } from '@angular/core';
import { NgxCanvasModule } from 'ngx-canvas';

@NgModule({
  imports: [
    NgxCanvasModule.forRoot()
  ]
})
export class AppModule {}
```


Now you can simply use the `ngx-canvas` component and provide your custom options via the `options` binding:

```typescript
import { Component, OnInit } from '@angular/core';
import { DrawProps, ImageProps, TextProps, PropagateProps, RectProps, LineProps } from 'ngx-canvas';

@Component({
  selector: 'app-root',
  template: `
    <ngx-canvas [options]="options" (drawComplete)="drawComplete($event)"></ngx-canvas>
  `,
})
export class AppComponent implements OnInit {
  options: DrawProps;

  ngOnInit(): void {
    // draw image
    const imgs: ImageProps[] = [
      {
        type: 'image',
        url: 'assets/images/bg.png',
        top: 0,
        left: 0,
        width: 375,
        height: 667
      }
    ];
    // draw text
    const txts: TextProps[] = [
      {
        type: 'text',
        content: 'Have good time.',
        top: 100,
        left: 100,
        fontSize: 18,
        color: '#FFFFFF',
        lineHeight: 30,
        width: 200,
        lineNum: 1,
        fontFamily: 'Microsoft YaHei'
      }
    ];
    // draw rect
    const rects: RectProps[] = [
      {
        type: 'rect',
        width: 375,
        height: 667,
        x: 0,
        y: 280,
        backgroundColor: '#FFFFFF',
        borderRadius: 30
      }
    ];
    // draw line
    const lines: LineProps[] = [
      {
        type: 'line',
        color: '#999',
        startX: 10,
        startY: 200,
        endX: 350,
        endY: 200,
        width: 2,
        lineCap: 'round'
      }
    ];
    // draw options
    this.options = {
      debug: true, // Used to debug the presentation canvas
      width: 375,
      height: 667,
      backgroundColor: '#47b785',
      views: [
        ...imgs,
        ...txts,
        ...rects,
        ...lines
      ],
      extra: 'This is an additional optional any parameter that will be returned unchanged after the drawing is completed.'
    };
  }

  drawComplete(propagate: PropagateProps): void {
    const { dataUrl, canvas, ctx, extra } = propagate;
    // downloading canvas element to an image
    const a = document.createElement('a');
    document.body.appendChild(a);
    a.setAttribute('style', 'display: none');
    a.href = dataUrl;
    a.download = 'filename';
    a.click();
    a.remove();
  }
}
```

Notice that you will need to import the `NgxCanvasModule` into other modules as it exports `ngx-canvas` component. But `forRoot` has to be called only once!


## API

### Bindings

| @Input()        | Type              | Required    | Description                        |
| --------------- | ----------------- | ----------- | ---------------------------------  |
| options         | `DrawProps`       | required    | Options of the canvas.             |
| width           | `number`          | optional    | Width of the canvas display.       |
| height          | `number`          | optional    | Height of the canvas display.      |


### Events

| @Output()        | Type                            | Required    | Description       |
| ---------------- | ------------------------------- | ----------- | ----------------- |
| drawComplete     | `(propagate: PropagateProps) => unknown`  | optional    | Gets an data that is notified when the canvas is finished draw.      |


## Types

| DrawProps        | Type              | Required    | Description                        | Default    |
| ---------------- | ----------------- | ----------- | ---------------------------------  | -----------|
| width            | `number`          | required    | Generate image width.              | null       |
| height           | `number`          | required    | Generated image height.            | null       |
| backgroundColor  | `string`          | optional    | Canvas background color.           | '#FFFFFF'  |
| debug            | `boolean`         | optional    | Whether to enable debugging mode.  | false      |
| views            | `Array<T>`        | required    | Canvas drawing core data.          | null       |
| extra            | `any`             | optional    | Some extra data.                   | null       |

---

| ImageProps       | Type              | Required    | Description                                         | Default              |
| ---------------- | ----------------- | ----------- | --------------------------------------------------- | -------------------- |
| type             | `string`          | required    | Type of drawing.                                    | Fixed value 'image'  |
| url              | `string`          | required    | Picture path or remote address.                     | null                 |
| top              | `number`          | required    | Y-axis coordinate in the destination canvas.        | 0                    |
| left             | `number`          | required    | X-axis coordinate in the destination canvas.        | 0                    |
| width            | `number`          | required    | Width to draw the image in the destination canvas.  | 0                    |
| height           | `number`          | required    | Height to draw the image in the destination canvas. | 0                    |
| borderRadius     | `number`          | optional    | Image borderRadius.                                 | 0                    |
| borderWidth      | `number`          | optional    | Image borderWidth.                                  | 0                    |
| borderColor      | `string`          | optional    | Image borderColor.                                  | 'rgba(255,255,255,0)'|

---

| TextProps        | Type              | Required    | Description                                         | Default              |
| ---------------- | ----------------- | ----------- | --------------------------------------------------- | -------------------- |
| type             | `string`          | required    | Type of drawing.                                    | Fixed value 'text'   |
| top              | `number`          | required    | Y-axis coordinate in the destination canvas.        | 0                    |
| left             | `number`          | required    | X-axis coordinate in the destination canvas.        | 0                    |
| content          | `string`          | required    | Text string to render into the context.             | null                 |
| fontSize         | `number`          | optional    | Font size.                                          | 16                   |
| baseLine         | `string`          | optional    | Current text baseline used when drawing text.       | '#000'               |
| textAlign        | `string`          | optional    | Current text alignment used when drawing text.      | 'top'                |
| opacity          | `number`          | optional    | Font transparency.                                  | 'left'               |
| width            | `number`          | optional    | Maximum text width.                                 | null                 |
| lineNum          | `number`          | optional    | Maximum Text lines.                                 | 1                    |
| lineHeight       | `number`          | optional    | Font line height.                                   | 0                    |
| fontWeight       | `number`          | optional    | Font weight.                                        | 'normal'             |
| fontStyle        | `string`          | optional    | Font style.                                         | 'normal'             |
| fontFamily       | `string`          | optional    | Font family.                                        | 'Microsoft YaHei'    |

--- 

| LineProps        | Type              | Required    | Description                                         | Default              |
| ---------------- | ----------------- | ----------- | --------------------------------------------------- | -------------------- |
| type             | `string`          | required    | Type of drawing.                                    | Fixed value 'line'   |
| startX           | `number`          | required    | X-axis coordinate of the line's start point.        | null                 |
| startY           | `number`          | required    | Y-axis coordinate of the line's start point.        | null                 |
| endX             | `number`          | required    | X-axis coordinate of the line's end point.          | null                 |
| endY             | `number`          | required    | Y-axis coordinate of the line's end point.          | null                 |
| color            | `string`          | optional    | Color of the line.                                  | '#000'               |
| width            | `number`          | optional    | Width of the line.                                  | 1                    |
| lineCap          | `string`          | optional    | Shape used to draw the end points of lines.         | 'butt'               |


---

| RectProps        | Type              | Required    | Description                                         | Default              |
| ---------------- | ----------------- | ----------- | --------------------------------------------------- | -------------------- |
| type             | `string`          | required    | Type of drawing.                                    | Fixed value 'rect'   |
| width            | `number`          | required    | Rectangle's width.                                  | 0                    |
| height           | `number`          | required    | Rectangle's height.                                 | null                 |
| x                | `number`          | required    | X-axis coordinate of the rectangle's start point.   | null                 |
| y                | `number`          | required    | Y-axis coordinate of the rectangle's start point.   | null                 |
| text             | `TextProps`       | optional    | Text string to render into the context.             | null                 |
| paddingLeft      | `number`          | optional    | Padding area to the left of rectangle's.            | 0                    |
| paddingRight     | `number`          | optional    | Padding area to the right of rectangle's.           | 0                    |
| borderWidth      | `number`          | optional    | Width of rectangle's border.                        | null                 |
| backgroundColor  | `string`          | optional    | BackgroundColor of rectangle's                      | null                 |
| borderColor      | `string`          | optional    | Shape used to draw the end points of lines.         | null                 |
| borderRadius     | `number`          | optional    | Radius of rectangle's border.                       | 0                    |
| opacity          | `number`          | optional    | Rect transparency.                                  | 1                    |

---

| PropagateProps   | Type                            | Description                                         |
| ---------------- | ------------------------------- | --------------------------------------------------- |
| canvas           | `HTMLCanvasElement`             | Canvas Instance Object.                             |
| ctx              | `CanvasRenderingContext2D`      | CanvasRenderingContext2D interface Object.          |
| dataUrl          | `string`                        | Data URI containing a representation of the image.  |
| extra            | `any`                           | Additional parameters are returned after drawing.   |


## Demo

You can clone this repo to your working copy and then launch the demo page in your local machine:

```vim
npm install
npm run build:prod
npm run start
```

The demo page server is listening on: http://localhost:4200

CHANGELOG.md
## [0.0.8](https://github.com/lanxuexing/ngx-canvas/compare/v0.0.6...v0.0.8) (2021-01-17)


### Features

* modify the return parameters of the finished drawing ([2c51b8d](https://github.com/lanxuexing/ngx-canvas/commit/2c51b8d50ed221d215754319844e4af0dadb1674))
* update example app ([4122e7b](https://github.com/lanxuexing/ngx-canvas/commit/4122e7b32247ad83bd76bbe9453a2badb5ae197b))
* update repos keywords ([01c28c1](https://github.com/lanxuexing/ngx-canvas/commit/01c28c1944805d1e3875e1f2b5107624bfc803bd))
* upgrade lib version ([0800004](https://github.com/lanxuexing/ngx-canvas/commit/0800004618d82dba7aee9d10f2761c71c6c2bce8))



## [0.0.6](https://github.com/lanxuexing/ngx-canvas/compare/7e1893d9cb643372b4bdd3c3c05c5929ac505d50...v0.0.6) (2021-01-11)


### Bug Fixes

* canvas debug options ([92f1e36](https://github.com/lanxuexing/ngx-canvas/commit/92f1e36f6cf855c871914074532ca3f93b141ba3))
* options init error ([01dfdc3](https://github.com/lanxuexing/ngx-canvas/commit/01dfdc3e83e7ac3b13f2d2ba38fe38c1ec4a4650))


### Features

* example ([09d857f](https://github.com/lanxuexing/ngx-canvas/commit/09d857f9d0ef3aa26d7c800044f8991935bdf82b))
* init project ([7e1893d](https://github.com/lanxuexing/ngx-canvas/commit/7e1893d9cb643372b4bdd3c3c05c5929ac505d50))





projects/ngx-canvas/src/lib/ngx-canvas.component.ts
import {
  AfterViewInit,
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnChanges,
  OnDestroy,
  Output,
  SimpleChanges,
  ViewChild
} from '@angular/core';
import {
  DrawProps,
  ImageProps,
  LineProps,
  PropagateProps,
  RectProps,
  TextProps
} from './type';

@Component({
  selector: 'ngx-canvas',
  template: `
    <canvas
      [hidden]="!options?.debug"
      [style.--width]="width"
      [style.--height]="height"
      #canvas>
    </canvas>
  `,
  styles: [`
    canvas {
      width: calc(var(--width) * 1px);
      height: calc(var(--height) * 1px);
    }
  `]
})
export class NgxCanvasComponent implements AfterViewInit, OnDestroy, OnChanges {
  @Input() options: DrawProps;
  @Input() width: number;
  @Input() height: number;
  @Output() drawComplete: EventEmitter<PropagateProps> = new EventEmitter();
  @ViewChild('canvas') canvasRef: ElementRef;

  public canvas: HTMLCanvasElement;
  public ctx: CanvasRenderingContext2D;
  private timer: number;

  constructor() { }

  ngAfterViewInit(): void {
    this.initCanvas();
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (!changes.options.firstChange) {
      this.initCanvas();
    }
  }

  /**
   * @description 初始化canvas画布
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/canvas
   */
  private initCanvas(): void {
    this.canvas = this.canvasRef.nativeElement;
    if (!this.options?.width || !this.options?.height || this.options?.views.length === 0) {
      return;
    }
    this.canvas.width = this.options.width;
    this.canvas.height = this.options.height;
    this.ctx = this.canvas.getContext('2d');
    this.drawArr();
  }


  /**
   * @description 绘图
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D
   */
  private async drawArr(): Promise<void> {
    // setps1：绘制背景色
    if (this.options.backgroundColor) {
      this.ctx.save();
      this.ctx.fillStyle = this.options.backgroundColor;
      this.ctx.fillRect(0, 0, this.options.width, this.options.height);
      this.ctx.restore();
    }
    for (const iterator of this.options.views) {
      if (iterator.type === 'image') {
        // setps2：绘制图片
        await this.drawImg(iterator);
      } else if (iterator.type === 'text') {
        // setps3：绘制文字
        this.drawText(iterator);
      } else if (iterator.type === 'rect') {
        // setps4：绘制矩形
        this.drawRect(iterator);
      } else if (iterator.type === 'line') {
        // steps5：绘制线段
        this.drawLine(iterator);
      }
    }
    // steps6：绘制完成向外发射数据和事件
    const propagate: PropagateProps = {
      canvas: this.canvas,
      ctx: this.ctx,
      dataUrl: this.canvas.toDataURL(),
      extra: this.options?.extra
    };
    this.drawComplete.emit(propagate);
  }

  /**
   * @description 绘制图片
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/drawImage
   * @param data 数据对象
   */
  drawImg(data: ImageProps): Promise<any> {
    return new Promise<any>((resolve, reject) => {
      const {
        url, top = 0, left = 0, width = 0, height = 0, borderRadius = 0, borderWidth = 0, borderColor = 'rgba(255,255,255,0)'
      } = data;
      this.ctx.save();
      const img = new Image();
      img.setAttribute('crossorigin', 'anonymous');
      img.src = url;
      img.onload = () => {
        if (borderRadius > 0) {
          this.drawRadiusRect(left, top, width, height, borderRadius, borderWidth, borderColor);
          this.ctx.clip();
          this.ctx.drawImage(img, left, top, width, height);
        } else {
          this.ctx.drawImage(img, left, top, width, height);
        }
        this.ctx.restore();
        this.timer = window.setTimeout(() => {
          resolve(true);
        }, 100);
      };
      img.onerror = (err) => {
        reject(err);
      };
    });
  }

  /**
   * @description 绘制圆角矩形
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/arc
   * @param x 距离左边的距离
   * @param y 距离顶部的距离
   * @param w 矩形的宽
   * @param h 矩形的高
   * @param r 矩形的圆角
   * @param borderWidth 矩形边框的宽
   * @param borderColor 矩形边框的颜色
   */
  drawRadiusRect(x: number, y: number, w: number, h: number, r: number, borderWidth?: number, borderColor?: string): void {
    const br = r / 2;
    this.ctx.beginPath();
    if (borderWidth > 0) { this.ctx.lineWidth = borderWidth; }
    this.ctx.strokeStyle = borderColor;
    this.ctx.moveTo(x + br, y); // 移动到左上角的点
    this.ctx.lineTo(x + w - br, y);
    this.ctx.arc(x + w - br, y + br, br, 2 * Math.PI * (3 / 4), 2 * Math.PI * (4 / 4));
    this.ctx.lineTo(x + w, y + h - br);
    this.ctx.arc(x + w - br, y + h - br, br, 0, 2 * Math.PI * (1 / 4));
    this.ctx.lineTo(x + br, y + h);
    this.ctx.arc(x + br, y + h - br, br, 2 * Math.PI * (1 / 4), 2 * Math.PI * (2 / 4));
    this.ctx.lineTo(x, y + br);
    this.ctx.arc(x + br, y + br, br, 2 * Math.PI * (2 / 4), 2 * Math.PI * (3 / 4));
    this.ctx.closePath();
    this.ctx.stroke();
  }

  /**
   * @description 绘制文字
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/fillText
   * @param data 数据对象
   */
  drawText(data: TextProps): void {
    const {
      top = 0, left = 0, fontSize = 16, color = '#000', baseLine = 'top', textAlign = 'left', content, opacity = 1,
      width, lineNum = 1, lineHeight = 0, fontWeight = 'normal', fontStyle = 'normal', fontFamily = 'Microsoft YaHei'
    } = data;
    this.ctx.save();
    this.ctx.beginPath();
    this.ctx.font = `${fontStyle} ${fontWeight} ${fontSize}px ${fontFamily}`;
    this.ctx.globalAlpha = opacity;
    this.ctx.textAlign = textAlign;
    this.ctx.textBaseline = baseLine;
    this.ctx.fillStyle = color;
    let textWidth = this.ctx.measureText(content).width;
    const textArr = [];
    if (textWidth > width) {
      let fillText = '';
      let line = 1;
      for (let i = 0; i <= content.length - 1; i += 1) { // 将文字转为数组
        fillText += content[i];
        if (this.ctx.measureText(fillText).width >= width) {
          if (line === lineNum) {
            if (i !== content.length - 1) {
              fillText = `${fillText.substring(0, fillText.length - 1)}...`;
            }
          }
          if (line <= lineNum) {
            textArr.push(fillText);
          }
          fillText = '';
          line += 1;
        } else if (line <= lineNum) {
          if (i === content.length - 1) {
            textArr.push(fillText);
          }
        }
      }
      textWidth = width;
    } else {
      textArr.push(content);
    }
    textArr.forEach((item, index) => {
      this.ctx.fillText(item, left, top + (lineHeight || fontSize) * index);
    });
    this.ctx.restore();
  }

  /**
   * @description 绘制线段
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/lineTo
   * @param data 数据对象
   */
  drawLine(data: LineProps): void {
    const {
      startX, startY, endX, endY, color = '#000', width = 1, lineCap = 'butt',
    } = data;
    this.ctx.save();
    this.ctx.beginPath();
    this.ctx.lineCap = lineCap;
    this.ctx.strokeStyle = color;
    this.ctx.lineWidth = width;
    this.ctx.moveTo(startX, startY);
    this.ctx.lineTo(endX, endY);
    this.ctx.stroke(); // 进行绘制
    this.ctx.closePath();
    this.ctx.restore();
  }

  /**
   * @description 绘制矩形
   * @links https://developer.mozilla.org/zh-CN/docs/Web/API/CanvasRenderingContext2D/fillRect
   * @param data 数据对象
   */
  drawRect(data: RectProps): void {
    const {
      text, width = 0, height, x, y, paddingLeft = 0, paddingRight = 0, borderWidth,
      backgroundColor, borderColor, borderRadius = 0, opacity = 1
    } = data;
    // 判断是否块内有文字
    let blockWidth = 0; // 块的宽度
    let textX = 0;
    let textY = 0;
    if (typeof text !== 'undefined') {
      // 如果有文字并且块的宽度小于文字宽度，块的宽度为 文字的宽度 + 内边距
      const textWidth = this.getTextWidth(typeof text.content === 'string' ? text : text.content);
      blockWidth = textWidth > width ? textWidth : width;
      blockWidth += paddingLeft + paddingLeft;
      const { textAlign = 'left' } = text;
      textY = height / 2 + y; // 文字的y轴坐标在块中线
      if (textAlign === 'left') {
        // 如果是右对齐，那x轴在块的最左边
        textX = x + paddingLeft;
      } else if (textAlign === 'center') {
        textX = blockWidth / 2 + x;
      } else {
        textX = x + blockWidth - paddingRight;
      }
    } else {
      blockWidth = width;
    }
    if (backgroundColor) {
      // 画面
      this.ctx.save();
      this.ctx.globalAlpha = opacity;
      this.ctx.fillStyle = backgroundColor;
      if (borderRadius > 0) {
        // 画圆角矩形
        this.drawRadiusRect(x, y, blockWidth, height, borderRadius);
        this.ctx.fill();
      } else {
        this.ctx.fillRect(x, y, blockWidth, height);
      }
      this.ctx.restore();
    }
    if (borderWidth) {
      // 画线
      this.ctx.save();
      this.ctx.globalAlpha = opacity;
      this.ctx.fillStyle = borderColor;
      this.ctx.lineWidth = borderWidth;
      if (borderRadius > 0) {
        // 画圆角矩形边框
        this.drawRadiusRect(x, y, blockWidth, height, borderRadius);
        this.ctx.stroke();
      } else {
        this.ctx.strokeRect(x, y, blockWidth, height);
      }
      this.ctx.restore();
    }
    if (text) {
      this.drawText(Object.assign(text, { x: textX, y: textY }));
    }
  }

  /**
   * @description 动态计算文本长度
   * @param draweText 数组 | 对象
   */
  private getTextWidth(draweText: Array<any> | object): number {
    let texts = [];
    if (Object.prototype.toString.call(draweText) === '[object Object]') {
      texts.push(draweText);
    } else {
      texts = draweText as any[];
    }
    let width = 0;
    texts.forEach(({ fontSize, text, marginLeft = 0, marginRight = 0 }) => {
      this.ctx.font = `${fontSize}`;
      width += this.ctx.measureText(text).width + marginLeft + marginRight;
    });
    return width;
  }

  ngOnDestroy(): void {
    if (this.timer) { clearTimeout(this.timer); }
  }

}
